File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        ans = []

        def backtrack(i: int, subset: List[int], total: int):
            if i >= len(subset) or total > target:
                return
            
            subset.append(nums[i])
            backtrack(i, subset, total + nums[i])
            subset.pop()
            backtrack(i + 1, subset, total - nums[i])

        backtrack(0, [], 0)

        return ans

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.js

class Solution {
	/**
	 * @param {number[]} candidates
	 * @param {number} target
	 * @return {number[][]}
	 */
	combinationSum2(candidates, target) {
		//let index = 0;
		let answer = [];
		let buffer = [];
		candidates.sort((a, b) => a - b);
		function backtrack(curTotal = 0, index = 0) {
            if (curTotal === target) {
				answer.push([...buffer]);
				return;
			}
			if (curTotal > target || index >= candidates.length) return;
			
			for (let i = index; i < candidates.length; i++) {

				buffer.push(candidates[i]);
				backtrack(curTotal + candidates[i], i + 1);
				buffer.pop();
			}
		}
		backtrack();
		return answer.filter((curArr, i, twoDarray) => JSON.stringify(curArr) === JSON.stringify(twoDarray[i + 1]) ? false : true);
	}

}
